pkg/tsdb/cloudwatch/cloudwatch.go

```diff
@@ -0,0 +1,381 @@
+package cloudwatch
+
+import (
+	"context"
+	"errors"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/grafana/grafana/pkg/log"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/tsdb"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/cloudwatch"
+	cwapi "github.com/grafana/grafana/pkg/api/cloudwatch"
+	"github.com/grafana/grafana/pkg/components/null"
+	"github.com/grafana/grafana/pkg/components/simplejson"
+	"github.com/grafana/grafana/pkg/metrics"
+)
+
+type CloudWatchExecutor struct {
+	*models.DataSource
+}
+
+func NewCloudWatchExecutor(dsInfo *models.DataSource) (tsdb.Executor, error) {
+	return &CloudWatchExecutor{
+		DataSource: dsInfo,
+	}, nil
+}
+
+var (
+	plog               log.Logger
+	standardStatistics map[string]bool
+	aliasFormat        *regexp.Regexp
+)
+
+func init() {
+	plog = log.New("tsdb.cloudwatch")
+	tsdb.RegisterExecutor("cloudwatch", NewCloudWatchExecutor)
+	standardStatistics = map[string]bool{
+		"Average":     true,
+		"Maximum":     true,
+		"Minimum":     true,
+		"Sum":         true,
+		"SampleCount": true,
+	}
+	aliasFormat = regexp.MustCompile(`\{\{\s*(.+?)\s*\}\}`)
+}
+
+func (e *CloudWatchExecutor) Execute(ctx context.Context, queries tsdb.QuerySlice, queryContext *tsdb.QueryContext) *tsdb.BatchResult {
+	var result *tsdb.BatchResult
+	queryType := queries[0].Model.Get("type").MustString()
+	switch queryType {
+	case "timeSeriesQuery":
+		result = e.executeTimeSeriesQuery(ctx, queries, queryContext)
+		break
+	case "metricFindQuery":
+		result = e.executeMetricFindQuery(ctx, queries, queryContext)
+		break
+	}
+	return result
+}
+
+func (e *CloudWatchExecutor) executeTimeSeriesQuery(ctx context.Context, queries tsdb.QuerySlice, queryContext *tsdb.QueryContext) *tsdb.BatchResult {
+	result := &tsdb.BatchResult{
+		QueryResults: make(map[string]*tsdb.QueryResult),
+	}
+
+	errCh := make(chan error, 1)
+	resCh := make(chan *tsdb.QueryResult, 1)
+
+	currentlyExecuting := 0
+	for _, model := range queries {
+		queryType := model.Model.Get("type").MustString()
+		if queryType != "timeSeriesQuery" {
+			continue
+		}
+		currentlyExecuting++
+		go func(refId string) {
+			queryRes, err := e.executeQuery(ctx, model.Model.Get("parameters"), queryContext)
```

Comment:
This might cause problems since model is a pointer to queries[i]. when the loop iterates model will be updated. 
In this case Get() will return a new value so it will work as expected but I think its better to be explicit. like the refId sent to this function.

Reply:
This is fixed by you. Thanks.